gen/gen.go

```diff
@@ -69,31 +84,45 @@ func Gen(exercise string, j interface{}, t *template.Template) error {
 	if dirMetadata == "" {
 		return errors.New("unable to determine current path")
 	}
+
 	jFile := filepath.Join("exercises", exercise, "canonical-data.json")
 	// find and read the json source file
-	jPath, jOrigin, jCommit := getPath(jFile)
-	jSrc, err := ioutil.ReadFile(filepath.Join(jPath, jFile))
-	if err != nil {
-		return err
+	jPath, jOrigin, jCommit := getLocal(jFile)
+	var body []byte
+	fullPath := filepath.Join(jPath, jFile)
+	_, err := os.Stat(fullPath)
+	switch {
+	case err != nil:
+		body, jOrigin, jCommit, err = getRemote(exercise)
+		if err != nil {
+			return err
+		}
+	case err == nil:
+		body, err = ioutil.ReadFile(fullPath)
+		if err != nil {
+			log.Fatal(err)
+		}
 	}
 
-	// unmarshal the json source to a Go structure
-	if err = json.Unmarshal(jSrc, j); err != nil {
-		// This error message is usually enough if the problem is a wrong
-		// data structure defined here. Sadly it doesn't locate the error well
-		// in the case of invalid JSON.  Use a real validator tool if you can't
-		// spot the problem right away.
-		return fmt.Errorf(`unexpected data structure: %v`, err)
-	}
-
-	// These fields are guaranteed to be in every problem
-	var commonMetadata struct {
-		Version string
+	// Decode the json
+	err = json.NewDecoder(bytes.NewReader(body)).Decode(&j)
```

Comment:
> it should probably stay as it originally was, using `json.Unmarshal`, no?

Right, doesn't seem to make much sense to go from `io.Reader` to `[]byte` back to `io.Reader`. Seems to make sense to keep it as `json.Unmarshal` or return an `io.Reader` from `getRemote` and `getLocal`. Probably good hygiene to close that `Reader` too (which implies it has to be a `ReadCloser`.